web/elm/src/Tooltip.elm

```diff
@@ -14,12 +16,28 @@ type alias Model m =
     { m | hovered : HoverState.HoverState }
 
 
-type TooltipPolicy
+type TooltipCondition
     = AlwaysShow
     | OnlyShowWhenOverflowing
 
 
-policy : DomID -> TooltipPolicy
+type alias AttachPosition =
+    { direction : Direction
+    , alignment : Alignment
+    }
+
+
+type Direction
+    = Top
+    | Right
+
+
+type Alignment
+    = Start
+    | Middle
```

Comment:
Cool approach! Just as food for thought, I had previously implemented anchor-based positioning where entities had 9 possible anchor-points:

```
x--x--x
|  |  |
x--x--x
|  |  |
x--x--x
```

And to position elements, you would anchor one point to another.

For instance, with the tooltip example, you'd anchor the `BottomLeft` anchor-point of the tooltip to the `TopLeft` anchor-point of the sync icon:

```
*---------*---------*
|         |         |
*---------*---------*
|         |         |
@--x--x---*---------*
|  |  |
x--x--x
|  |  |
x--x--x
```

Each point was defined as an `x_0`, `y_0`, and `anchor` (where `x_0` and `y_0` are the coordinates of the top-left of the entity), and each anchor had a tuple of percentages along the `x` and `y` directions (e.g. `BottomMiddle` was `(0.5, 1.0)`). Then, computing the actual coordinates of a point (relative to the dimensions of its entity) is easy:

```
x = x_0 + width * x_pct
y = y_0 + height * y_pct
```

Anchoring a point to another becomes a matter of computing `x` and `y` above for the parent point, and producing a new point where `x_0 = x`, `y_0 = y`, and `anchor` is set appropriately,

My use-case required more flexibility than we probably need with tooltips: a configuration like "anchor `BottomLeft` to `TopRight`" isn't likely to be useful.

```
      *---------*---------*
      |         |         |
      *---------*---------*
      |         |         |
x--x--@---------*---------*
|  |  |
x--x--x
|  |  |
x--x--x
```

Your way might also be more intuitive, but figured I'd provide an alternative idea!

Reply:
when you say "had previously implemented" - do you mean elsewhere in this codebase, or on another project? I had toyed with an API similar to what you've described, but as you suggest, I found it a bit overkill. The "anchoring" formalism as you've described seems strictly more expressive than the attach/align one, so we could even implement attach/align in terms of anchors. If there's precedent.

Reply:
I meant on another project so you won't be able to reuse it here